### PR TITLE
Add design/api.md

### PR DESCRIPTION
**Pull request by wangkuiyi**

### CODE REVIEW

File: doc/design/api.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+```python
+gm = paddle.gradient_machine.new(model)  # gm uses default cost function.
+ud = paddle.updater.new_simple()  # A simple updater doesn't work with parameter servers.
+rd = paddle.data.amazon.product_review.new()
+mt = paddle.metric.new_acc()
+
+for mb, pass_end in rd.read():
+    gm.feed(mb)
+    ud.update(gm)
```
Comment (jacquesqiao) [inline doc/design/api.md]: 这里应该需要 gm.forward()  gm.backward()

Comment (wangkuiyi) [inline doc/design/api.md]: 哦。我这里忘了说明白——gm.feed调用了forward和 backward。

gm.update是什么意思呢？是更新模型参数吗？我以为是 ud.update 来做这个事儿的。

File: doc/design/api.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+hidden_gen = paddle.layer.fc(input_gen, ...)
+output_gen = paddle.layer.fc(hidden_gen, ...)
+
+# gm_gen maintains layer outputs and gradients of model gen.
+gm_gen = paddle.gradient_machine.new(output_gen)
+
+input_dis = paddle.layer.input(...)
+hidden_dis = paddle.layer.fc(intput_dis, ...)
+output_dis = paddle.layer.softmax(hidden_dis, ...)
+
```
Comment (emailweixu) [inline doc/design/api.md]: For GAN, gm_dis and gm_gen update different subset of parameters. We need to have a mechanism to specify this. In the current GAN example, this is achieved by setting is_static according to is_discriminator_training. This is possible in the current GAN example because the configs of gm_gen and gm_dis are actually generated differently.

File: doc/design/api.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+To write a Paddle program using the current API, we would have to write two Python source files -- one defines the data provider and the other defines the network or run the for loop.  This doesn't work well with Notebooks.  So we decide to redesign the API.  This document describes the basic design concerns.
+
+## Basic Concepts
+
+The API design depends on basic concepts about deep learning.
+
+### Model
+
+For deep learning, a model includes two parts: the topology and parameters.  Currently, the concept *model* in Paddle contains only the topology, and parameters are in another concept *gradient machine*.  This differs from the intuition and makes it difficult to save/load models.  In this design, we should keep both topology and parameters in a *model*.
```
Comment (helinwang) [inline doc/design/api.md]: Here is how tensorflow do it for reference:
* Tensorflow has a concept of graph (topology), load graph: https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go#Graph.Import

* Save and load weights is a Operation on graph, save and load in golang: https://gist.github.com/helinwang/7782c6b2815c334c77653fc0e52b6069

File: doc/design/api.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+ud = paddle.updater.new_simple()
+
+rd = paddle.data.mnist.new()
+
+for mb, pass_end in rd.read():
+    fake = gm_gen.forward(mb)
+    fake_label = paddle.input.new(False, ...)
+    real_label = paddle.input.new(True, ...)
+    gm_dis.feed([fake, fake_label])
+    gm_dis.feed([mb, real_label])
```
Comment (helinwang) [inline doc/design/api.md]: Maybe this question is naive, will the computed gradient from second `feed` override the gradients from first `feed`?

File: doc/design/api.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    ud.update(gm_dis)
+    
+    gm_comp.feed([mb, real_label])
```
Comment (helinwang) [inline doc/design/api.md]: Since updater will only need to update gradients of `output_gen`'s predecessors'. One possible optimization here is to let gradient machine know this information. So it does not need to save any unrelated activation and gradients. E.g.,
```
gm_comp.feed([mb, real_label], output_gen)
ud.update(gm_comp, output_gen)
```

Maybe I don't have enough context, but from a first glance, I feel maybe it's easier that gradient machine be stateless (current design saves gradients inside gradient machine). E.g.,
```
// pseudo code
type gradientMachine
type gradient
type updater

var gm_comp gradientMachine
var ud updater
var gradient g = gm_comp.feed([mb, real_label], output_gen)
ud.update(g)
```

File: doc/design/api.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    ud.update(gm_comp, output_gen) # updates only the model whose output layer is output_gen.
+```
+
+A key point here is that we use the output layer to indicate a model.  I think that we can achieve this as long as each layer knows about its predecessors so that we can trace from the output layer upward till the input layers.  Please be aware that we didn't compose two models in above example code; instead, we only created a gradient machine which covers both `model_gen` and `model_dis`.
```
Comment (helinwang) [inline doc/design/api.md]: Just for reference: If I understand correctly, tensorflow do it by explicitly state weights that needs update.
```
fine_tune_step = tf.train.AdamOptimizer(1e-2).minimize(cross_entropy, var_list=[W_fc2, b_fc2])
fine_tune_step.run(feed_dict={x: batch[0], y_: batch[1], keep_prob: 0.5})
```

File: doc/design/api.md (hunk 2 above)
Comment (wangkuiyi) [inline doc/design/api.md]: 我理解有这个需求。我这个设计里是这样实现的：

1. 每个“部分”描述成一个“网络”

1. 每个“网络”用其output指代。

   比如例子里有 `output_dis` 和 `output_gen`。这可以通过在每个layer中记录其所有input layers来实现，这样给定一个output layer，我们即可trace到整个网络中的所有layers。

1. 每个“网络”的更新信息维护在一个 gradient machine。

   例子中，有两个网络，但是却有三个gradient machines：gm_dis 对应 output_dis，gm_gen 对应 output_gen，gm_comp 对应 output_gen 和 output_dis 的组合。

1. 对每个网络的更新，是通过 updater 来实现的。

   updater的输入是 gradient machine，因为通过每次forward/backward 调用计算得到的 layer outputs 和 gradients 都保存在 gradient machine 里了，而且通过 gradient machine 可以查到其对应的network的信息。

   在下面例子里：`ud.update(gm_dis)` 是利用 gm_dis 更新其对应的 output_dis 网络；而 `ud.update(gm_comp, output_gen)` 是利用 gm_comp 更新其对应的网络中的 `output_gen` 子网络。

File: doc/design/api.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+### A Simple Network
+
+```python
+gm = paddle.gradient_machine.new(model)  # gm uses default cost function.
+ud = paddle.updater.new_simple()  # A simple updater doesn't work with parameter servers.
+rd = paddle.data.amazon.product_review.new()
+mt = paddle.metric.new_acc()
+
+for mb, pass_end in rd.read():
```
Comment (wangkuiyi) [inline doc/design/api.md]: 确实。可以通过read函数指定mini batch size，也可以在创建reader的时候指定，比如

```
rd = paddle.data.amazon.product_review.new(minibatch_size=100)
```

File: doc/design/api.md (hunk 4 above)
Comment (wangkuiyi) [inline doc/design/api.md]: 这个设计里是这么假设的：第二次 feed 调用产生的 layer outputs 会覆盖之前一次 feed调用产生的；第二次 feed 调用产生的gradients 也会覆盖之前一次调用产生的。

File: doc/design/api.md (hunk 5 above)
Comment (wangkuiyi) [inline doc/design/api.md]: 这里可能是我没有说清楚，所以引起误会了。`gm_comp` 这个 gradient machine 对应的是 gen 和 dis 两个网络的组合。这个组合的输入是 gen 的输入，所以 input 只能是 mb。

File: doc/design/api.md (hunk 6 above)
Comment (wangkuiyi) [inline doc/design/api.md]: 理解了。在这个设计里，也有类似的描述：

```
ud.update(gm_dis)
```

利用 gm_dis 更新其对应的 output_dis 网络，其中没有显示指定需要被更新的部分，所以是更新gm_dis 对应的整个网络；而

```
ud.update(gm_comp, output_gen)
```

是利用 gm_comp 更新其对应的网络中的 `output_gen` 子网络。其中 `output_gen` 是显示指定的需要被更新的部分。

File: doc/design/api.md (hunk 3 above)
Comment (wangkuiyi) [inline doc/design/api.md]: 这一点我没有特别理解，为什么graph和weights要分开呢？按照model是graph加上parameterss（weights）的概念，两者貌似放在一个class里更自然呀？

File: doc/design/api.md (hunk 1 above)
Comment (jacquesqiao) [inline doc/design/api.md]: 确实应该是 ud.update(gm) 而不是 gm.update，我说的不对~

File: doc/design/api.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+   In order to learn the model, we need to run the error backpropagation algorithm iteratively.  In each iteration, we run the forward algorithm with a minibatch of data.  This updates the outputs of layers.  Then we run a backward algorithm which computes the gradients of every parameter.  The outputs and gradients are not part of the model; instead, they are side effects of the training process and should be maintained in the trainer.
+   
+   After the invocation of the backward algorithm, we should update model parameters using the gradients and parameters like learning rate.  This step might involve communications with the parameter server in the case of distributed training.  This complexity motivates us to separate the trainer into two concepts:
+   
+   1. *gradient machine*, which computes and maintains layer outputs and gradients, and
+      
+   1. *updater*, which encapsulates the updating algorithm.
+   
+   It seems that *cost function* is a property of *gradient machine*?
```
Comment (wangkuiyi) [inline doc/design/api.md]: cost function和模型无关，只和训练方法有关，不应该是网络的一部分。

File: doc/design/api.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+   
+1. *Data*
+
+   Models are trained using data sets.  We hope to provide a set of utility data sets encapsulated in Python packages like `paddle.data.amazon.product_review` and `paddle.data.wikipedia.articles`.  A reasonable idea might be that in each of these packages, we provide a `new` function that returns a reader object or a Python iterator.  And the *reader* has a read method `read`, which, once called, returns a minibatch and a flag indicating if it reaches the end of a data set.  For online learning, the flag would always be False.  Therefore, a training loop might look like:
```
Comment (wangkuiyi) [inline doc/design/api.md]: API的设计应该和语言无关。应该使用各种语言都有的功能，而不专门依赖Python语法。

File: doc/design/api.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+## Examples
+
+### A Simple Network
+
+```python
+gm = paddle.gradient_machine.new(model)  # gm uses default cost function.
```
Comment (wangkuiyi) [inline doc/design/api.md]: 在我写的正文里，没有强调，但是在我给徐老师和鹤麟的回复里有强调：

1. 没有model这个概念了等价于network。所以应该既没有model，也没有model creator的概念才对。
1. 每个network用它的output layer指代。具体请参见 https://github.com/PaddlePaddle/Paddle/pull/1088#discussion_r95085035

**Comment by reyoung** [comment]
目前的分歧包括

## 神经网络模型是不是包括loss function?

一方认为神经网络模型不包括loss function。原因

* 模型预测的时候其实是不用loss function的。比如分类问题，可以用交叉熵，也可以用huber loss之类的其他loss。但是模型预测并不需要知道这些loss是什么。
* 训练的时候，直接用 model.fit(output, label, cost="cross-entropy") 

另一方认为神经网络模型应该加上loss function。原因

* 有一些模型的loss function特别复杂，如果分开写的话很困难
* 有一些无监督学习，loss function并不能在最后指定，也是模型的一部分。
   * Neural Art模型，中的content_loss和style_loss是两组网络的对比。如果在最后指定的话，可能比较复杂。 https://github.com/jcjohnson/neural-style/blob/master/neural_style.lua#L108

## Paddle模型配置应该怎么定义?

1. 函数

  ```python
  @define_network(inputs={'pixel': dense_vector(784)})
  def network(pixel):
    hidden = fc_layer(input=pixel, size=200)
    pred = fc_layer(input=hidden, size=10, act=SoftmaxActivation())
    return pred
  ```

2. 网络的最后一个变量

  ```python
  pixel = data_layer(name='pixel', type=dense_vector(784))
  hidden = fc_layer(input=pixel, size=200)
  pred = fc_layer(input=hidden, size=10, act=SoftmaxActivation())  # pred store all neural network topology
  ```

问题是，第二种变成变量很难实现，因为那样基本上要把Paddle目前的网络解析重构一遍了。
 
## 网络的表示是否提供了足够灵活性，能描述GAN这样的问题

分歧是，GAN这种网络，是需要『网络描述』时的灵活性，还是训练时的灵活性？GAN基本问题是在更新的时候选择更新参数。这其实是updater的update函数，需要在更新的时候，指定一个[Predicate](http://en.cppreference.com/w/cpp/concept/Predicate) 吧。与网络描述似乎无关。

File: doc/design/api.md (hunk 9 above)
Comment (reyoung) [inline doc/design/api.md]: python iterator只是一种类的调用约定。本质上是一个对象，但是这个对象实现了 `next` 方法。调用next的时候，返回一组数据。当iterator空了的时候，throw一个StopIteration的exception。
这和*reader*概念类似，next函数也就相当于read函数。Python的iterator就相当于java里面的List interface一样自然。

如果我们封装的是 "Python"的api，在不增加用户心智负担的情况下，接口还是尽量按照"Python"的逻辑实现吧。

File: doc/design/api.md (hunk 10 above)
Comment (OleNet) [inline doc/design/api.md]: 同意， decorator给用户带来了极大的困扰
